lib/galaxy/webapps/galaxy/api/workflows.py

```diff
@@ -980,6 +983,253 @@ def _generate_report(self, trans, invocation_id, **kwd):
+                }
+                input_subdomain.append(input_obj)
+
+        usability_domain = []
+        for a in stored_workflow.annotations:
+            usability_domain.append(a.annotation)
+        for h in history.annotations:
+            usability_domain.append(h.annotation)
+
+        parametric_domain = []
+        for inv_step in workflow_invocation.steps:
+            try:
+                for k, v in inv_step.workflow_step.tool_inputs.iteritems():
+                    param, value, step = k, v, inv_step.workflow_step.order_index
+                    parametric_domain.append({'param': param, 'value': value, 'step': step})
+            except Exception:
+                continue
+
+        execution_domain = {
+            'script_access_type': 'a_galaxy_workflow',
+            'script': [url_for('workflows', encoded_workflow_id=encoded_workflow_id)],
+            'script_driver': 'Galaxy',
+            'software_prerequisites': software_prerequisites,
+            'external_data_endpoints': [],
+            'environment_variables': {}
+        }
+
+        extension = [
+            {
+                'extension_schema': 'https://raw.githubusercontent.com/biocompute-objects/extension_domain/6d2cd8482e6075746984662edcf78b57d3d38065/galaxy/galaxy_extension.json',
+                'galaxy_extension': {
+                    'galaxy_url': url_for('/', qualified=True),
+                    'galaxy_version': VERSION
+                }
+            }
+        ]
+
+        error_domain = {
+            'empirical_error': kwd.get('empirical_error', []),
+            'algorithmic_error': kwd.get('algorithmic_error', [])
+        }
+
+        bco_dict = {
+            'bco_id': url_for('invocation_export_bco', invocation_id=invocation_id, qualified=True),
+            'spec_version': spec_version,
+            'etag': str(model.uuid4().hex),
+            'provenance_domain': provenance_domain,
+            'usability_domain': usability_domain,
+            'extension_domain': extension,
+            'description_domain': {
+                'keywords': keywords,
+                'xref': kwd.get('xref', []),
+                'platform': ['Galaxy'],
+                'pipeline_steps': pipeline_steps,
+            },
+            'execution_domain': execution_domain,
+            'parametric_domain': parametric_domain,
+            'io_domain': {
+                'input_subdomain': input_subdomain,
+                'output_subdomain': output_subdomain,
+            },
+            'error_domain': error_domain,
+        }
+        return bco_dict
+
+    @expose_api
+    def download_invocation_bco(self, trans, invocation_id, **kwd):
+        """
+        GET /api/invocations/{invocations_id}/get_bco
+
+        Returns a selected BioCompute Object.
+
+        """
+        ret_dict = self.export_invocation_bco(trans, invocation_id, **kwd)
+
+        return format_return_as_json(ret_dict)
```

Comment:
Maybe something like this works?

```suggestion
        trans.response.headers["Content-Disposition"] = 'attachment; filename="bco_%s.json"' % invocation_id
        return ret_dict
```

Reply:
Thanks @nsoranzo! It works but the downloaded file still has the `\` for escaping the quotations. Any idea about getting rid of that?

Reply:
I did not know, but I've found how to fix it! `@expose_api_raw` is the correct decorator for file downloads.

Also calling `export_invocation_bco()` inside `download_invocation_bco()` would give you a JSON object because of the `@expose_api` decorator of the former, so I had to split the generation of bco into an helper method which returns a simple dict. In this way we can call `format_return_as_json(ret_dict, pretty=True)` to produce a pretty-printed JSON to download.